File: S02/week14/sejun/MajorityElement.java
// 7ms, 49.68mb
class Solution {
		class Element {
			int count;
			int value;

			public Element(int count, int value) {
				this.count = count;
				this.value = value;
			}

			public void inc() {
				count++;
			}

			public int getCount() {
				return count;
			}

			public int getValue() {
				return value;
			}
		}
		Map<Integer, Element> elementMap = new HashMap<>();

		public int majorityElement(int[] nums) {

			int majorityElementMinCount = nums.length / 2 + nums.length % 2 ;
			Arrays.stream(nums).forEach(num -> {
				if (elementMap.containsKey(num)) {
					elementMap.get(num).inc();
				} else {
					elementMap.put(num, new Element(1, num));
				}
			});


			for (Integer value : elementMap.keySet()) {
				if (elementMap.get(value).count >= majorityElementMinCount) {
					return value;
				}
			}

			return -1;
		}
	}

File: S02/week14/seokjoon/MajorityElement.kt
//237ms , 40.38mb
class Solution {
    fun majorityElement(nums: IntArray): Int {
        val map = mutableMapOf<Int,Int>()
        val targetCount = nums.size / 2 + 1

        nums.forEach {
            map[it] = map.getOrDefault(it,0) + 1
            if(map.getValue(it) >= targetCount) return it
        }
        // 만약 여기까지 도달한다면.. 
        var result = 0
        map.forEach { key, value ->
            if(value >= targetCount) result = key
        }

        return result
    }
}
